pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,556 @@
+				//Something changed.
+				glog.V(4).Infof("Something changed for service %q: stopping it", serviceName)
+				proxier.removeServiceInternal(serviceName, info)
+			}
+
+			glog.V(1).Infof("Adding new service %q at %s:%d/%s", serviceName, serviceIP, servicePort.Port, servicePort.Protocol)
+			info = proxier.addServiceInternal(serviceName)
+			info.portal.ip = serviceIP
+			info.portal.port = servicePort.Port
+			info.portal.protocol = servicePort.Protocol
+			info.deprecatedPublicIPs = service.Spec.DeprecatedPublicIPs
+			// Deep-copy in case the service instance changes
+			info.loadBalancerStatus = *api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer)
+			info.sessionAffinityType = service.Spec.SessionAffinity
+
+			glog.V(4).Infof("info: %+v", info)
+		}
+	}
+
+	for name, info := range proxier.serviceMap {
+		if !activeServices[name] {
+			glog.V(1).Infof("Removing service %q", name)
+			proxier.removeServiceInternal(name, info)
+		}
+	}
+
+	glog.V(6).Infof("Syncing rules.")
+	if err := proxier.iptablesInit(); err != nil {
+		glog.Errorf("Failed to ensure iptables: %v", err)
+	}
+	// this locks proxier.mu, hence the closure above
+	if err := proxier.syncProxyRules(); err != nil {
+		glog.Errorf("Failed to sync iptables rules: %v", err)
+	}
+}
+
+// OnEndpointsUpdate takes in a slice of updated endpoints.
+func (proxier *Proxier) OnEndpointsUpdate(allEndpoints []api.Endpoints) {
+	registeredEndpoints := make(map[proxy.ServicePortName]bool)
+	proxier.mu.Lock()
+	defer proxier.mu.Unlock()
+
+	// Update endpoints for services.
+	for i := range allEndpoints {
+		svcEndpoints := &allEndpoints[i]
+
+		// We need to build a map of portname -> all ip:ports for that
+		// portname.  Explode Endpoints.Subsets[*] into this structure.
+		portsToEndpoints := map[string][]hostPortPair{}
+		for i := range svcEndpoints.Subsets {
+			ss := &svcEndpoints.Subsets[i]
+			for i := range ss.Ports {
+				port := &ss.Ports[i]
+				for i := range ss.Addresses {
+					addr := &ss.Addresses[i]
+					portsToEndpoints[port.Name] = append(portsToEndpoints[port.Name], hostPortPair{addr.IP, port.Port})
+				}
+			}
+		}
+
+		for portname := range portsToEndpoints {
+			svcPort := proxy.ServicePortName{types.NamespacedName{svcEndpoints.Namespace, svcEndpoints.Name}, portname}
+			state, exists := proxier.serviceMap[svcPort]
+			if !exists || state == nil {
+				state = proxier.addServiceInternal(svcPort)
+			}
+			curEndpoints := []string{}
+			if state != nil {
+				curEndpoints = state.endpoints
+			}
+			newEndpoints := flattenValidEndpoints(portsToEndpoints[portname])
+
+			if len(curEndpoints) != len(newEndpoints) || !slicesEquiv(slice.CopyStrings(curEndpoints), newEndpoints) {
+				glog.V(1).Infof("EndpointsTracker: Setting endpoints for %s to %+v", svcPort, newEndpoints)
+				state.endpoints = newEndpoints
+			}
+			registeredEndpoints[svcPort] = true
+		}
+	}
+	// Remove endpoints missing from the update.
+	for service := range proxier.serviceMap {
```

Comment:
It might simplify lifecycle if you only delete from the serviceMap when the Service ceases to exist.  A Service with no Endpoints is actually not a big deal.

Actually, now I am concerned about out-of-order delivery.  Maybe you should keep the serviceMap entry if either an Endpoints or Service object exists.

Consider what happens in Endpoints arrive first: we add the serviceMap entry.  Now we get an update of Services without that Service.  We delete the serviceMap entry.  Now we get the Service - create.  Now we have missed the Endpoints, and we will end up deleting the Service again.

Instead: Insert the serviceMap entry if the Service exists in an Update or if the Endpoints exists in an Update.  When a Service Update comes in without a given Service, check whether its Endpoints is a nil slice.  If so delete it.  If not, leave it.  We'll eventually catch up.

If we want to be smarter, mark it as expired, but keep it around until Endpoints object is removed, then delete it if expired.